src/main/scala/mesosphere/marathon/core/health/impl/HealthCheckActor.scala

```diff
@@ -252,11 +252,9 @@ private[health] class HealthCheckActor(
     // dead task (in addition to the living one), thus sometimes leading
     // Marathon to report the task / instance / app as unhealthy while
     // everything is running correctly.
-    val healthOfInstanceId = healthByTaskId.find(_._1.instanceId == instanceId)
-    if (healthOfInstanceId.isDefined)
-      healthByTaskId.remove(healthOfInstanceId.get._1)
+    healthIndex.remove(taskId)
```

Comment:
This part should still be fixed, as the behavior is a little bit different than before, and it was precisely done on purpose to fix a bug.
InstanceID are assigned for instances (as seen on marathon), but Marathon can try to start the instance several times (thus the different TaskIDs).
If the health checks are received out of order, there is no way to know which is an old health check (thus to be thrown away), and which is the new one (and should be kept).
The implemented patch was to find the healthcheck belonging to the same isntanceID to delete it, so as the new one always takes place, which will eventually be replaced when the new Task will be running for good.

In your current patch, I see we remove by instanceID, but I'm not sure we would removed other TaskID with the same instanceID ?

Reply:
OK, as seen in the meeting, I put back the old behaviour.